airbyte-integrations/connectors/source-mailgun/source_mailgun/source.py

```diff
@@ -0,0 +1,154 @@
+            Mapping[str, Any]:
+        return {}
+
+
+class Events(IncrementalMailgunStream):
+    # TODO: Event Polling. See https://documentation.mailgun.com/en/latest/api-events.html#event-polling
+
+    cursor_field = "timestamp"
+
+    primary_key = "id"
+
+    def __init__(self, config: Mapping[str, Any], *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.timestamp = config.get('timestamp', 0)
+        self.ascending = config.get('ascending', True)
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> \
+            Mapping[str, Any]:
+        if current_stream_state is not None and 'timestamp' in current_stream_state:
+            current_parsed_timestamp = current_stream_state['timestamp']
+            latest_record_timestamp = latest_record['timestamp']
+            return {'timestamp': max(current_parsed_timestamp, latest_record_timestamp)}
+        else:
+            return {'timestamp': self.timestamp}
+
+    def path(self, *args, next_page_token: Optional[Mapping[str, Any]] = None, **kwargs) -> str:
+        # return super().path(*args, **kwargs) or "events"  # TODO: Requires the CDK update (see MailgunStream.path())
+        path: str = "events"
+        if next_page_token:
+            token: str = next_page_token['url'].rpartition('/')[2]
+            path = f"{path}/{token}"
+
+        return path
+
+    def request_params(
+            self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None,
+            next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state, stream_slice, next_page_token)
+        params.update({
+            "begin": stream_state.get("timestamp", self.timestamp),
+            "ascending": "yes" if self.ascending else "no",
+        })
+        return params
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        """
+        Override default stream_slices CDK method to provide date_slices as a list of a single slice with such format:
+        [{"timestamp": 1636411500.861427}]
+        """
+        return [stream_state]
+
+
+# Source
+class SourceMailgun(AbstractSource):
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        """
+        :param config:  the user-input config object conforming to the connector's spec.json
+        :param logger:  logger object
+        :return Tuple[bool, any]: (True, None) if the input config can be used to connect to the API successfully, (False, error) otherwise.
+        """
+        try:
+            response = requests.get(urljoin(MailgunStream.url_base, 'domains'), auth=('api', config['private_key']))
+            if response.status_code == 200:
+                return True, None
+            else:
+                return False, response.json()['message']
```

Comment:
no guarantee that response is in json format (500, html/text)

Reply:
also this method doesn't retry requests in case of retriable failure, that is why in most places we use streams